lib/media/segment_prefetch.js

```diff
@@ -0,0 +1,173 @@
+/*! @license
+ * Shaka Player
+ * Copyright 2016 Google LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+goog.require('shaka.net.NetworkingEngine');
+goog.require('shaka.media.InitSegmentReference');
+goog.require('shaka.media.SegmentReference');
+goog.provide('shaka.media.SegmentPrefetch');
+goog.require('shaka.log');
+
+/**
+ * @summary
+ * This class manages segment prefetch operations.
+ * Called by StreamingEngine to prefetch next N segments
+ * ahead of playhead, to reduce the chances of rebuffering.
+ */
+shaka.media.SegmentPrefetch = class {
+  /**
+   * @param {number} prefetchLimit
+   * @param {shaka.extern.Stream} stream
+   * @param {function(
+   *  !(shaka.media.InitSegmentReference|shaka.media.SegmentReference),
+   *  shaka.extern.Stream
+   * ):!shaka.net.NetworkingEngine.PendingRequest} fetchDispatcher
+   */
+  constructor(prefetchLimit, stream, fetchDispatcher) {
+    /**
+     * @private {number}
+     */
+    this.prefetchLimit_ = prefetchLimit;
+
+    /** @private {shaka.extern.Stream} */
+    this.stream_ = stream;
+
+    /** @private {number} */
+    this.prefetchPosTime_ = 0;
+
+    /** @private {function(
+     * !(shaka.media.InitSegmentReference|shaka.media.SegmentReference),
+     * shaka.extern.Stream
+     * ):!shaka.net.NetworkingEngine.PendingRequest} */
+    this.fetchDispatcher_ = fetchDispatcher;
+
+    /**
+     * @private {!Map.<shaka.media.SegmentReference,
+                 !shaka.net.NetworkingEngine.PendingRequest>}
+     */
+    this.segmentPrefetchMap_ = new Map();
+  }
+
+  /**
+   * Fetch next segments ahead of current segment.
+   *
+   * @param {(!shaka.media.SegmentReference)} startReference
+   * @public
+   */
+  prefetchSegments(startReference) {
+    if (!(this.prefetchLimit_ > 0)) {
+      return;
+    }
+    const logPrefix = shaka.media.SegmentPrefetch.logPrefix_(
+        this.stream_,
+    );
+    if (!this.stream_.segmentIndex) {
+      shaka.log.info(
+          logPrefix, 'missing segmentIndex',
+      );
+      return;
+    }
+    const currTime = startReference.startTime;
+    const maxTime = Math.max(currTime, this.prefetchPosTime_);
+    const iterator = this.stream_.segmentIndex.getIteratorForTime(maxTime);
+    let reference = startReference;
+    while (this.segmentPrefetchMap_.size < this.prefetchLimit_ &&
+            reference != null) {
+      if (!this.segmentPrefetchMap_.has(reference)) {
+        const op = this.fetchDispatcher_(reference, this.stream_);
+        this.segmentPrefetchMap_.set(reference, op);
+      }
+      this.prefetchPosTime_ = reference.startTime;
+      reference =iterator.next().value;
+    }
+  }
+
+  /**
+   * Get the result of prefetched segment if already exists.
+   * @param {(
+   *  !shaka.media.InitSegmentReference|!shaka.media.SegmentReference
+   * )} reference
+   * @return {?shaka.net.NetworkingEngine.PendingRequest} op
+   * @public
+   */
+  getPrefetchedSegment(reference) {
+    let op = null;
+    if (
```

Comment:
I assume that returning null here if the input is not a SegmentReference is to leave open the possibility of fetching init segments at a later date?
If so, it should probably be a separate return clause, with a comment explaining that it's a placeholder.
That would also remove the need for this multi-line if-statement.

Reply:
actually it was a mistake to leave InitSegmentReferece as an input type, so i changed the input type to only be SegmentReference